xla/service/gpu/tests/gpu_fused_mha_test.cc

```diff
@@ -1387,6 +1387,110 @@ class FlashAttentionPagedAttention : public MultiHeadedAttentionTest {
+      %Arg_1.2 = bf16[4,1024,4,64]{3,2,1,0} parameter(1)
+      %transpose.1 = bf16[4,4,64,1024]{3,2,1,0} transpose(%Arg_1.2), dimensions={0,2,3,1}
+      %reshape.13 = bf16[16,64,1024]{2,1,0} reshape(%transpose.1)
+      %dot.3 = f32[16,1024,1024]{2,1,0} dot(%reshape.12, %reshape.13), lhs_batch_dims={0}, lhs_contracting_dims={2}, rhs_batch_dims={0}, rhs_contracting_dims={1}
+      %reshape.14 = f32[4,4,1024,1024]{3,2,1,0} reshape(%dot.3)
+      %constant = f32[] constant(0.333333343)
+      %broadcast = f32[4,4,1024,1024]{3,2,1,0} broadcast(%constant), dimensions={}
+      %multiply = f32[4,4,1024,1024]{3,2,1,0} multiply(%reshape.14, %broadcast)
+      %tanh.12 = f32[4,4,1024,1024]{3,2,1,0} tanh(%multiply)
+      %constant.6 = f32[] constant(3)
+      %broadcast.7 = f32[4,4,1024,1024]{3,2,1,0} broadcast(%constant.6), dimensions={}
+      %multiply.13 = f32[4,4,1024,1024]{3,2,1,0} multiply(%tanh.12, %broadcast.7)
+      %constant.9 = f32[] constant(-inf)
+      %reduce.18 = f32[4,4,1024]{2,1,0} reduce(%multiply.13, %constant.9), dimensions={3}, to_apply=%region_0.14
+      %broadcast.23 = f32[4,4,1024,1024]{3,2,1,0} broadcast(%reduce.18), dimensions={0,1,2}
+      %subtract.24 = f32[4,4,1024,1024]{3,2,1,0} subtract(%multiply.13, %broadcast.23)
+      %exponential.25 = f32[4,4,1024,1024]{3,2,1,0} exponential(%subtract.24)
+      %constant.8 = f32[] constant(0)
+      %reduce.30 = f32[4,4,1024]{2,1,0} reduce(%exponential.25, %constant.8), dimensions={3}, to_apply=%region_1.26
+      %broadcast.34 = f32[4,4,1024,1024]{3,2,1,0} broadcast(%reduce.30), dimensions={0,1,2}
+      %divide.35 = f32[4,4,1024,1024]{3,2,1,0} divide(%exponential.25, %broadcast.34)
+      %convert.36 = bf16[4,4,1024,1024]{3,2,1,0} convert(%divide.35)
+      %reshape.16 = bf16[16,1024,1024]{2,1,0} reshape(%convert.36)
+      %dot.4 = f32[16,64,1024]{2,1,0} dot(%reshape.15, %reshape.16), lhs_batch_dims={0}, lhs_contracting_dims={2}, rhs_batch_dims={0}, rhs_contracting_dims={2}
+      %reshape.17 = f32[4,4,64,1024]{3,2,1,0} reshape(%dot.4)
+      %convert.0 = bf16[4,4,64,1024]{3,2,1,0} convert(%reshape.17)
+      ROOT %transpose.10 = bf16[4,1024,4,64]{3,2,1,0} transpose(%convert.0), dimensions={0,3,1,2}
+    }
+  )";
+    return hlo_text;
+  }
+
+  template <typename T>
+  void TestImpl_Flash_Attention_Flex_Attention() {
+    if (skip_reason_) GTEST_SKIP() << *skip_reason_;
+    if (GetDnnVersionInfoOrDefault(backend().default_stream_executor()) <
+        se::dnn::VersionInfo(9, 7, 0)) {
+      GTEST_SKIP() << "Flash Attention requires cuDNN >= 9.7.0.";
+    }
+    // Extend cudnn sdpa soft capping using flex attention
+    std::string hlo_string =
+        GetModuleFlash_Attention_Flex_Attention_Soft_Capping_BF16();
+    // Reference implementation is pure xla sdpa soft capping.
+    std::string hlo_string_ref =
+        GetModuleFlash_Attention_Flex_Attention_Soft_Capping_Reference_BF16();
+    EXPECT_TRUE(RunAndCompareTwoModules(hlo_string, hlo_string_ref,
```

Comment:
This code makes sure the custom call computes the same result as the HLO above, right?
Should we test the code that rewrites the HLO into the custom call too?

Reply:
There is no rewrite happening in XLA for cudnn sdpa anymore. It is comparing the cudnn custom call against pure xla implementation.

Reply:
Thanks for the explanation. I would argue that JAX is a better place for this test than XLA.

The idea is to ensure that cuDNN sdpa is equivalent to JAX's default HLO lowering, right? Let's add this test to JAX, so we cover the JAX HLO lowering too (which currently isn't tested). XLA codegen will be covered either way.